datalad_next/patches/push_to_export_remote.py

```diff
@@ -0,0 +1,240 @@
+            **credential_properties) or [(None, None)])[0][1]
+    return credentials
+
+
+# TODO: create a function similar to AnnexRepo.get_special_remote() and put it
+#  into core.
+def get_export_records(repo: AnnexRepo) -> Generator:
+    """Read exports that git-annex recorded in its 'export.log'-file
+
+    Interpret the lines in export.log. Each line has the following structure:
+
+        time-stamp " " source-annex-uuid ":" destination-annex-uuid " " treeish
+
+    Parameters
+    ----------
+    repo: AnnexRepo
+        The annex repo from which exports should be determined
+
+    Returns
+    -------
+    List[Dict]
+        Dictionary containing the keys: "timestamp", "source-annex-uuid",
+        "destination-annex-uuid", "treeish". The timestamp-value is a float,
+        all other values are strings.
+    """
+    try:
+        for line in repo.call_git_items_(["cat-file", "blob", "git-annex:export.log"]):
+            result_dict = dict(zip(
+                [
+                    "timestamp",
+                    "source-annex-uuid",
+                    "destination-annex-uuid",
+                    "treeish"
+                ],
+                line.replace(":", " ").split()))
+            result_dict["timestamp"] = float(result_dict["timestamp"][:-1])
+            yield result_dict
+    except CommandError as command_error:
+        # Some errors indicate that there was no export yet.
+        expected_errors = (
+            "fatal: Not a valid object name git-annex:export.log",
+        )
+        if command_error.stderr.strip() in expected_errors:
+            return []
+        raise
+
+
+def _get_export_log_entry(repo: AnnexRepo,
+                          target_uuid: str
+                          ) -> Optional[Dict]:
+    target_entries = [
+        entry
+        for entry in get_export_records(repo)
+        if entry["destination-annex-uuid"] == target_uuid]
+
+    if not target_entries:
+        return None
+    return sorted(target_entries, key=lambda e: e["timestamp"])[-1]
+
+
+def _is_valid_treeish(repo: AnnexRepo,
+                      export_entry: Dict,
+                      ) -> bool:
+
+    for line in repo.call_git_items_(["log", "--pretty=%H %T"]):
+        commit_hash, treeish = line.split()
+        if treeish == export_entry["treeish"]:
+            return True
+    return False
+
+
+def _transfer_data(repo: AnnexRepo,
+                   ds: Dataset,
+                   target: str,
+                   content: Iterable,
+                   data: str,
+                   force: Optional[str],
+                   jobs: Optional[Union[str, int]],
+                   res_kwargs: Dict,
+                   got_path_arg: bool
+                   ) -> Generator:
+
+    target_uuid, remote_info = ([
+        (uuid, info) for uuid, info in repo.get_special_remotes().items()
+        if info.get("name") == target] or [(None, None)])[0]
+
+    if _is_export_remote(remote_info):
```

Comment:
Personal style note: In such cases (almost the entire function body is a conditional), i find it to be cognitively less complex to reverse the conditional, and to the minimal `else` first, followed by an explicit return. Like so:

```suggestion
    if not _is_export_remote(remote_info):
            yield from push._push_data(
            ds,
            target,
            content,
            data,
            force,
            jobs,
            res_kwargs.copy(),
            got_path_arg=got_path_arg,
        )
        return
```

and then take the main conditional and deindent it. Essentially saying "we don't handle this one thing here", rather than saying "here is a list of all things we handle here". Otherwise there are 60 lines to scroll to remember what was the other thing.

Or put the conditional clause in a function, whatever makes the main decision immediately obvious when going through the function from start to end.

Reply:
Addressed as suggested